### DISCUSSION TITLE
Creating my own delegate on macOS

### DISCUSSION THREAD

**Discussion by MarkSetchell** [ImageMagick/ImageMagick#6914] [Help]
### ImageMagick version

7.1

### Operating system, version and so on

macOS Sonoma and all previous

### Description

The homebrew version of ImageMagick no longer supports X11 in any variant (even if you have Xquartz installed), so you cannot do `magick -size 640x480 xc:red display:`. I therefore tried to make a delegate that uses `Preview` which is a native Apple app that will display an image. So, I made a file called `$HOME/.config/ImageMagick/delegates.xml` containing:

`<delegate decode="miff" encode="display" spawn="True" command="magick %i %u.png && /usr/bin/open -a Preview %u.png"/>`

And it works really well, I can now do:

`magick -size 640x480 xc:red display:`

and it displays fine. However, if I do:

`magick wizard: display:`

I get 2 wizards side-by-side. 

<img width="706" alt="Screenshot 2023-12-04 at 12 37 38" src="https://github.com/ImageMagick/ImageMagick/assets/12937410/d79daf2f-76a9-45c8-aa4b-e7474e84ddcf">

And if I do:

`magick rose: display:`

I get a rose on the left and a similarly sized black image to its right (alpha channel?). 

If I change the `delegates.xml` to PNG instead of MIFF, like this:

`<delegate decode="png" encode="display" spawn="True" command="magick %i %u.png && /usr/bin/open -a Preview %u.png"/>`

everything works. I don't understand this. It is not related to using macOS Preview app, because it does exactly the same if I just write the file to the local directory from the delegate with something like:

`<delegate decode="miff" encode="display" spawn="True" command="magick %i result.jpg"/>`

Any help please?

**Comment by fmw42** [comment]
Good question, Mark!  I would like to know the answer, also.

**Reply by jasonp019** [reply]
And, is there any way to have delegates just work for the display command?

This strikes me a sorcery without a good explanation for how it works to enable us to adapt it for general use.

It also strikes me as a *VERY* poor decision by the homebrew folks to disable linking to the X11 libraries.

**Reply by fmw42** [reply]
Mark,  have you tried just launching XQuartz first, then use the normal magick display command rather than create a new delegate.  I do that on my Mac, though I install IM from source and get my delegates from MacPorts.
